### PR TITLE
Add global banner manager to Wagtail

### PR DESCRIPTION
**Pull request by Scotchester**
We have created a lot of banners that live above everything else, at the very top of the page, with a [complex](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/jinja2/v1/_includes/organisms/header.html#L28-L61) [flag-based](https://github.com/cfpb/cfgov-refresh/blob/6de3110eac363de2969043fbe7c8ccb51f50632b/cfgov/cfgov/settings/base.py#L631-L656) [apparatus](https://github.com/search?q=repo%3Acfpb%2Fcfgov-refresh+extension%3Ahtml+filename%3Abanner+language%3AHTML+language%3AHTML&type=Code&ref=advsearch&l=HTML&l=HTML). This requires developer intervention for every new banner that gets created, and it is piling on a lot of conditional code and tech debt that could better be handled by storing some of this configuration in the database.

This PR creates a new Wagtail-based banner manager to do just that. A new Banner Django model is added, along with a ModelAdmin interface for creating and editing Banner instances.

Banners have a `title` used for internal reference, a `url_patterns` field for determining what paths it appears on, an instance of the `Notification` molecule (providing the visible content of the notification), and an `enabled` checkbox to manually enable/disable the banner.

See: https://GHE/CFGOV/platform/issues/3677

## Additions

- `Banner` model in the `v1` app
- `banners` context variable on `CFGOVPage` for templates to check to see if any banners apply to the given request
- Template code in `organisms/header.html` to check said `banners` variable and output any banners that are in it

## Testing

1. Pull branch
1. Migrate
1. Go to http://localhost:8000/admin/v1/banner/ and create a new banner
1. Check that it appears on the pages you expect, and doesn't on the pages you don't expect

## Notes

- I don't immediately see why my additions to the `CFGOVPage` `get_context` function are causing errors like this with existing tests:

```
======================================================================
ERROR: test_get_context_calls_get_context (v1.tests.models.test_base.TestCFGOVPage)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/Users/cranfills/Projects/cfgov-refresh/cfgov/v1/tests/models/test_base.py", line 32, in test_get_context_calls_get_context
    self.page.get_context(self.request)
  File "/Users/cranfills/Projects/cfgov-refresh/cfgov/v1/models/base.py", line 214, in get_context
    enabled_banners = Banner.objects.filter(enabled=True)
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/query.py", line 787, in filter
    return self._filter_or_exclude(False, *args, **kwargs)
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/query.py", line 805, in _filter_or_exclude
    clone.query.add_q(Q(*args, **kwargs))
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1250, in add_q
    clause, _ = self._add_q(q_object, self.used_aliases)
  File "/Users/cranfills/Projects/cfgov-refresh/.tox/unittest-current/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1260, in _add_q
    connector = q_object.connector
AttributeError: 'Q' object has no attribute 'connector'
```

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### CODE REVIEW

File: cfgov/v1/models/banners.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+from v1.atomic_elements.molecules import Notification
+
+
+class Banner(models.Model):
+    title = models.CharField(
+        max_length=255,
+        help_text='For internal reference only; does not appear on the site.'
+    )
+    url_patterns = models.CharField(
```
Comment (chosak) [inline cfgov/v1/models/banners.py]: Nitpicky, but why is this plural? Shouldn't it be `url_pattern`?

File: cfgov/v1/models/banners.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        max_length=1000,
+        verbose_name='URL patterns',
+        help_text=mark_safe('A regular expression pattern for matching URLs '
+                            'that should show the banner, for example: '
+                            '<code>contact-us|^/complaint/$</code>'),
+        validators=[
+            RegexValidator(regex=r'[A-Za-z0-9\-_.:/?&|\^$]')
```
Comment (chosak) [inline cfgov/v1/models/banners.py]: So this is a regex that validates another regex?

File: cfgov/v1/models/banners.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        ],
+    )
+    # TODO: add `min_num` and `max_num` arguments of 1 to the StreamField below
+    # if https://github.com/wagtail/wagtail/pull/5185 ever gets merged.
```
Comment (chosak) [inline cfgov/v1/models/banners.py]: You can do this with a custom `StreamBlock`, for example as we do [here](https://github.com/cfpb/cfgov-refresh/blob/1b478eb9fe2b61ea29a344e898028da4f92130eb/cfgov/v1/models/sublanding_filterable_page.py#L29), but unfortunately it only works on the backend, at validation time, and won't prevent you from adding additional blocks in the editor. Wagtail 5185 has this same limitation, though.

For example:

```py
class BannerContent(StreamBlock):
    content = molecules.Notification()

    class Meta:
        block_counts = {
            'content': {'max_num': 1},
        }
```

File: cfgov/v1/models/banners.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    content = StreamField([
+        ('content', Notification()),
+    ])
+    enabled = models.BooleanField()
+
+    panels = [
+        FieldPanel('title'),
+        FieldPanel('url_patterns'),
```
Comment (chosak) [inline cfgov/v1/models/banners.py]: Worth using a [RegexField](https://docs.djangoproject.com/en/2.2/ref/forms/fields/#regexfield) in the editor?

```py
FieldPanel('url_patterns', widget=forms.RegexField(regex=REGEX_PATTERN))
```

File: cfgov/v1/models/base.py [hunk 5]
```diff
@@ -203,8 +206,16 @@ def get_appropriate_siblings(self, inclusive=True):
 
     def get_context(self, request, *args, **kwargs):
         context = super(CFGOVPage, self).get_context(request, *args, **kwargs)
+
         for hook in hooks.get_hooks('cfgovpage_context_handlers'):
             hook(self, request, context, *args, **kwargs)
+
+        context['banners'] = []
+        enabled_banners = Banner.objects.filter(enabled=True)
```
Comment (chosak) [inline cfgov/v1/models/base.py]: It's possible to simplify and optimize this further to do the regex matching as part of the database query:

```py
# Add any banners that are enabled and match the current request path.
context['banners'] = Banner.objects \
    .filter(enabled=True) \
    .annotate(path=Value(
        request.path,
        output_field=models.CharField()
    )) \
    .filter(path__regex=F('url_patterns'))
```

This creates an SQL query like this:

```sql
SELECT
    "v1_banner"."id",
    "v1_banner"."title",
    "v1_banner"."url_patterns",
    "v1_banner"."content",
    "v1_banner"."enabled",
    /about-us/careers/ AS "path"
FROM
    "v1_banner"
WHERE (
    "v1_banner"."enabled" = True AND
    /about-us/careers/::text ~ ("v1_banner"."url_patterns")
)
```

Here a banner has been defined with a regex `/about-us/`, and I'm visiting the page http://localhost:8000/about-us/careers/.

File: cfgov/jinja2/v1/_includes/organisms/header.html [hunk 6]
```diff
@@ -40,6 +40,17 @@
     </div>
     {% endif %}
 
+    {% if banners %}
+        {% for banner in banners %}
+            <div class="m-global-banner">
+                <div class="wrapper
+                          wrapper__match-content">
+                    {% include_block banner %}
```
Comment (chosak) [inline cfgov/jinja2/v1/_includes/organisms/header.html]:
```suggestion
                    {% include_block banner.content %}
```

This needs to be `banner.content` to include the `content` StreamField as opposed to the Django model instance.

File: cfgov/v1/models/banners.py (hunk 1 above)
Comment (Scotchester) [reply cfgov/v1/models/banners.py]: Yeah, I suppose you're right! Will update.

File: cfgov/v1/models/banners.py (hunk 2 above)
Comment (Scotchester) [reply cfgov/v1/models/banners.py]: Yep. It's regexes all the way down :)

I just entered the possible URL-safe characters off-the-cuff. Notice any I'm missing?

File: cfgov/v1/models/banners.py (hunk 3 above)
Comment (Scotchester) [reply cfgov/v1/models/banners.py]: Yeah, it just didn't seem worth adding the extra UI complication (adding the StreamBlock, then adding the Notification).

File: cfgov/v1/models/banners.py (hunk 4 above)
Comment (Scotchester) [reply cfgov/v1/models/banners.py]: 🤔 I'm not sure. Didn't know that was an option. I'll try it out.

File: cfgov/v1/models/base.py (hunk 5 above)
Comment (Scotchester) [reply cfgov/v1/models/base.py]: This seems like a great optimization! I will incorporate. Thanks!

File: cfgov/jinja2/v1/_includes/organisms/header.html (hunk 6 above)
Comment (Scotchester) [reply cfgov/jinja2/v1/_includes/organisms/header.html]: Yeah, with the way you suggested reworking the query, that's true.

File: cfgov/v1/models/banners.py (hunk 4 above)
Comment (Scotchester) [reply cfgov/v1/models/banners.py]: Doesn't work.

```
File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/core/handlers/exception.py" in inner
  41.             response = get_response(request)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/core/handlers/base.py" in _legacy_get_response
  249.             response = self._get_response(request)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  187.                 response = self.process_exception_by_middleware(e, request)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  185.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/decorators/cache.py" in _wrapped_view_func
  57.         response = view_func(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/urls/__init__.py" in wrapper
  102.             return view_func(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/decorators.py" in decorated_view
  34.             return view_func(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/options.py" in create_view
  326.         return view_class.as_view(**kwargs)(request)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/generic/base.py" in view
  68.             return self.dispatch(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/views.py" in dispatch
  691.         return super().dispatch(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/utils/decorators.py" in _wrapper
  67.             return bound_func(*args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/contrib/auth/decorators.py" in _wrapped_view
  23.                 return view_func(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/utils/decorators.py" in bound_func
  63.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/views.py" in dispatch
  79.         return super().dispatch(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/generic/base.py" in dispatch
  88.         return handler(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/generic/base.py" in get
  155.         context = self.get_context_data(**kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/views.py" in get_context_data
  150.         form = self.get_form()

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/generic/edit.py" in get_form
  44.             form_class = self.get_form_class()

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/views.py" in get_form_class
  127.         return self.get_edit_handler().get_form_class()

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/edit_handlers.py" in get_form_class
  344.             widgets=self.widget_overrides())

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/edit_handlers.py" in get_form_for_model
  62.     return metaclass(class_name, (form_class,), form_class_attrs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/forms/models.py" in __new__
  66.         new_class = super(WagtailAdminModelFormMetaclass, cls).__new__(cls, name, bases, attrs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/modelcluster/forms.py" in __new__
  245.         new_class = super(ClusterFormMetaclass, cls).__new__(cls, name, bases, attrs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/models.py" in __new__
  266.                 apply_limit_choices_to=False,

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/models.py" in fields_for_model
  186.             formfield = formfield_callback(f, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/forms/models.py" in formfield_for_dbfield
  47.     return db_field.formfield(**kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/db/models/fields/__init__.py" in formfield
  1110.         return super(CharField, self).formfield(**defaults)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/db/models/fields/__init__.py" in formfield
  891.         return form_class(**defaults)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/fields.py" in __init__
  228.         super(CharField, self).__init__(*args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/fields.py" in __init__
  109.         extra_attrs = self.widget_attrs(widget)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/fields.py" in widget_attrs
  246.         if self.max_length is not None and not widget.is_hidden:

Exception Type: AttributeError at /admin/v1/banner/create/
Exception Value: 'RegexField' object has no attribute 'is_hidden'
```

Comment (chosak) [reply cfgov/v1/models/banners.py]: You're right, I was mixing up widgets and form fields.

File: cfgov/v1/models/banners.py (hunk 3 above)
Comment (chosak) [reply cfgov/v1/models/banners.py]: This doesn't add any extra UI complication -- the Wagtail editing experience is the same. I missed one piece in my suggestion, which is the way to define the field:

```py
class Banner(models.Model):
    content = StreamField(BannerContent)
```

This avoids needing to add another top-level block.

File: cfgov/jinja2/v1/_includes/organisms/header.html (hunk 6 above)
Comment (chosak) [reply cfgov/jinja2/v1/_includes/organisms/header.html]: Oh, yes, you're right. I was testing against your branch with that local change applied.

File: cfgov/v1/models/banners.py (hunk 3 above)
Comment (Scotchester) [reply cfgov/v1/models/banners.py]: Ah, good call.

File: cfgov/v1/models/banners.py (hunk 1 above)
Comment (Scotchester) [reply cfgov/v1/models/banners.py]: Amended original commit with this change, now in fc9e180.

File: cfgov/v1/models/banners.py (hunk 3 above)
Comment (Scotchester) [reply cfgov/v1/models/banners.py]: Amended original commit with this change, now in fc9e180.

**Comment by Scotchester** [comment]
Travis is Travising, but the tests all pass for me locally, so I'm going unrequire the Travis check and merge this.

**Pull request merged**